### PR TITLE
fix a bug in `from_pretrained` when load optional components

### PR DESCRIPTION
**Pull request by yiyixuxu**
Below script currently fails but will be fixed with this PR (see more details about the error here  https://github.com/huggingface/diffusers/issues/4686
) 
```python
import torch
from diffusers import StableDiffusionUpscalePipeline

model_id = "stabilityai/stable-diffusion-x4-upscaler"
cache_path = "/home/yiyi_huggingface_co/cache"

pipe = StableDiffusionUpscalePipeline.from_pretrained(
    model_id,
    variant="fp16",
    torch_dtype=torch.float16
)
pipe.save_pretrained(cache_path)

pipe = StableDiffusionUpscalePipeline.from_pretrained(
    pretrained_model_name_or_path=cache_path,
    variant="fp16",
    torch_dtype=torch.float16
)
```

currently, when we call `from_pretrained()`,  optional components are treated same as optional kwargs - they are passed to the `init()` directly, but they should be handled as modules 

https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pipeline_utils.py#L1016

### CODE REVIEW

**Review by patrickvonplaten — APPROVED** [review]
Great find & fix! 

You're totally right. The problem is that for newer pipelines, we started allowing optional components to be `None`, e.g. as done here: https://github.com/huggingface/diffusers/blob/0a0fe69aa6a11b7723e83ca9e049e6096839ad4d/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_upscale.py#L102  but we didn't allow this in the beginning: https://github.com/huggingface/diffusers/blob/0a0fe69aa6a11b7723e83ca9e049e6096839ad4d/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L113

Your fix is very nice - could you please add a test for this? Maybe both for StableDiffusionUpscale and [StableDiffusionXL pipelines ](https://github.com/huggingface/diffusers/blob/0a0fe69aa6a11b7723e83ca9e049e6096839ad4d/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L131) (both have optional componens)

File: src/diffusers/pipelines/pipeline_utils.py [hunk 1]
```diff
@@ -1013,7 +1013,11 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
         init_dict, unused_kwargs, _ = pipeline_class.extract_init_dict(config_dict, **kwargs)
 
         # define init kwargs
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/pipeline_utils.py:1015]:
```suggestion
        # define init kwargs and make sure that optional component modules are filtered out
```

File: tests/pipelines/stable_diffusion_xl/test_stable_diffusion_xl.py [hunk 2]
```diff
@@ -689,3 +690,26 @@ def test_stable_diffusion_xl_multi_prompts(self):
 
         # ensure the results are not equal
         assert np.abs(image_slice_1.flatten() - image_slice_3.flatten()).max() > 1e-4
+
+    @require_torch_gpu
```
Comment (patrickvonplaten) [inline tests/pipelines/stable_diffusion_xl/test_stable_diffusion_xl.py:694]: Why do we need a GPU here - this should also work on CPU no?

File: tests/pipelines/stable_diffusion_2/test_stable_diffusion_upscale.py [hunk 3]
```diff
@@ -422,3 +423,42 @@ def test_stable_diffusion_pipeline_with_sequential_cpu_offloading(self):
         mem_bytes = torch.cuda.max_memory_allocated()
         # make sure that less than 2.9 GB is allocated
         assert mem_bytes < 2.9 * 10**9
+
+    def test_stable_diffusion_upscale_pipeline_from_save_pretrained(self):
```
Comment (patrickvonplaten) [inline tests/pipelines/stable_diffusion_2/test_stable_diffusion_upscale.py:427]: Can we use a dummy model here to not put too much burden on the tests (cc @DN6)

**Review by patrickvonplaten — APPROVED** [review]
Cool good to merge! 
@sayakpaul can you maybe also give this a review?

**Review by sayakpaul — APPROVED** [review]
Important one. Thank you.

**Pull request merged**